x/sync/manager.go

```diff
@@ -354,46 +384,267 @@ func (m *Manager) getAndApplyRangeProof(ctx context.Context, work *workItem) {
 		return
 	}
 
-	proof, err := m.config.Client.GetRangeProof(ctx,
-		&pb.SyncGetRangeProofRequest{
-			RootHash: targetRootID[:],
-			StartKey: &pb.MaybeBytes{
-				Value:     work.start.Value(),
-				IsNothing: work.start.IsNothing(),
-			},
-			EndKey: &pb.MaybeBytes{
-				Value:     work.end.Value(),
-				IsNothing: work.end.IsNothing(),
-			},
-			KeyLimit:   defaultRequestKeyLimit,
-			BytesLimit: defaultRequestByteSizeLimit,
+	request := &pb.SyncGetRangeProofRequest{
+		RootHash: targetRootID[:],
+		StartKey: &pb.MaybeBytes{
+			Value:     work.start.Value(),
+			IsNothing: work.start.IsNothing(),
 		},
-	)
+		EndKey: &pb.MaybeBytes{
+			Value:     work.end.Value(),
+			IsNothing: work.end.IsNothing(),
+		},
+		KeyLimit:   defaultRequestKeyLimit,
+		BytesLimit: defaultRequestByteSizeLimit,
+	}
+
+	requestBytes, err := proto.Marshal(request)
 	if err != nil {
+		m.finishWorkItem()
+		m.setError(err)
+		return
+	}
+
+	onResponse := func(ctx context.Context, _ ids.NodeID, responseBytes []byte, appErr error) {
+		defer m.finishWorkItem()
+
+		if err := m.handleRangeProofResponse(ctx, targetRootID, work, request, responseBytes, appErr); err != nil {
+			m.config.Log.Debug("dropping response", zap.Error(err))
+			m.retryWork(work)
+			return
+		}
+	}
+
+	if err := m.sendRequest(ctx, m.config.RangeProofClient, requestBytes, onResponse); err != nil {
+		m.finishWorkItem()
 		m.setError(err)
 		return
 	}
 
+	m.metrics.RequestMade()
+}
+
+func (m *Manager) sendRequest(ctx context.Context, client Client, requestBytes []byte, onResponse p2p.AppResponseCallback) error {
+	if len(m.config.StateSyncNodes) == 0 {
+		return client.AppRequestAny(ctx, requestBytes, onResponse)
+	}
+
+	// Get the next nodeID to query using the [nodeIdx] offset.
+	// If we're out of nodes, loop back to 0.
+	// We do this try to query a different node each time if possible.
+	nodeIdx := atomic.AddUint32(&m.stateSyncNodeIdx, 1)
+	nodeID := m.config.StateSyncNodes[nodeIdx%uint32(len(m.config.StateSyncNodes))]
+	return client.AppRequest(ctx, set.Of(nodeID), requestBytes, onResponse)
+}
+
+func (m *Manager) retryWork(work *workItem) {
```

Comment:
Another idea is to have a separate failed requests data structure. Feel like this will introduce more complexity into this PR though so for now we are just retrying by putting stuff in the work heap w/ it's own priority.

Reply:
https://github.com/ava-labs/avalanchego/blob/5c070f8ccefcf8ef5b7c77d17da88bccd05298fc/x/sync/client.go#L316

Previously we would make an outgoing request and if it failed, we would just retry. Since this was called synchronously it meant that the the blocking thread would basically just keep trying a request until it succeeded.

In the async style we're just failing and re-inserting it back onto the pending request heap with the highest priority. I think there's probably better ways to handle the retry strategy but I am trying to change it in this PR because this PR is already unmanageably big.